File: src/src_core/fmg/core/img/AMosaicsImg.cs
using System;
using System.Linq;
using System.Collections.Generic;
using System.ComponentModel;
using fmg.common.geom;
using fmg.common.geom.util;
using fmg.core.types;
using fmg.core.mosaic;
using fmg.core.mosaic.cells;

namespace fmg.core.img {

   /// <summary> Abstract representable <see cref="EMosaic"/> as image </summary>
   /// <typeparam name="TImage">plaform specific image</typeparam>
   public abstract class AMosaicsImg<TImage> : RotatedImg<TImage>, IMosaic
      where TImage : class
   {
      protected AMosaicsImg(EMosaic mosaicType, Matrisize sizeField) {
         _mosaicType = mosaicType;
         _sizeField = sizeField;
      }

      public enum ERotateMode {
         FullMatrix,
         SomeCells
      }

      private EMosaic _mosaicType;
      /// <summary>из каких фигур состоит мозаика поля</summary>
      public EMosaic MosaicType {
         get { return _mosaicType; }
         set {
            if (SetProperty(ref _mosaicType, value)) {
               Area = 0;
               _matrix.Clear();
               CellAttr = null;
            }
         }
      }

      private Matrisize _sizeField;
      public Matrisize SizeField {
         get { return _sizeField; }
         set {
            if (SetProperty(ref _sizeField, value)) {
               RecalcArea();
               _matrix.Clear();
               Invalidate();
            }
         }
      }

      public BaseCell getCell(Coord coord) { return Matrix[coord.x * SizeField.n + coord.y]; }

      private BaseCell.BaseAttribute _cellAttr;
      public BaseCell.BaseAttribute CellAttr {
         get {
            if (_cellAttr == null)
               CellAttr = MosaicHelper.CreateAttributeInstance(MosaicType); // call this setter
            return _cellAttr;
         }
         private set {
            if (SetProperty(ref _cellAttr, value)) {
               Dependency_CellAttribute_Area();
               Invalidate();
            }
         }
      }

      /// <summary> caching rotated values </summary>
      private readonly List<BaseCell> _matrix = new List<BaseCell>();
      /// <summary>матрица ячеек, представленная(развёрнута) в виде вектора</summary>
      public IList<BaseCell> Matrix {
         get {
            if (!_matrix.Any()) {
               var attr = CellAttr;
               var type = MosaicType;
               var size = SizeField;
               for (var i = 0; i < size.m; i++)
                  for (var j = 0; j < size.n; j++)
                     _matrix.Add(MosaicHelper.CreateCellInstance(attr, type, new Coord(i, j)));
               OnSelfPropertyChanged();
               Invalidate();
            }
            return _matrix;
         }
      }

      private void RecalcArea() {
         var w = Size.Width;
         var h = Size.Height;
         var pad = Padding;
         var sizeImageIn = new Size(w - pad.LeftAndRight, h - pad.TopAndBottom);
         var sizeImageOut = new SizeDouble(sizeImageIn.Width, sizeImageIn.Height);
         var area = MosaicHelper.FindAreaBySize(MosaicType, SizeField, ref sizeImageOut);
         Area = area; // call setter
         System.Diagnostics.Debug.Assert(w >= (sizeImageOut.Width + pad.LeftAndRight));
         System.Diagnostics.Debug.Assert(h >= (sizeImageOut.Height + pad.TopAndBottom));
         var paddingOut = new BoundDouble(
            (w - sizeImageOut.Width) / 2,
            (h - sizeImageOut.Height) / 2,
            (w - sizeImageOut.Width) / 2,
            (h - sizeImageOut.Height) / 2);
         System.Diagnostics.Debug.Assert((sizeImageOut.Width + paddingOut.LeftAndRight).HasMinDiff(w));
         System.Diagnostics.Debug.Assert((sizeImageOut.Height + paddingOut.TopAndBottom).HasMinDiff(h));

         PaddingFull = paddingOut;
      }

      private double _area;
      public double Area {
         get {
            if (_area <= 0)
               RecalcArea();
            return _area;
         }
         set {
            if (SetProperty(ref _area, value)) {
               Dependency_CellAttribute_Area();
               Invalidate();
            }
         }
      }

      private BoundDouble _paddingFull;
      public BoundDouble PaddingFull {
         get { return _paddingFull; }
         protected set {
            if (SetProperty(ref _paddingFull, value)) {
               Invalidate();
            }
         }
      }

      private ERotateMode _rotateMode = ERotateMode.FullMatrix;
      public ERotateMode RotateMode {
         get { return _rotateMode; }
         set { SetProperty(ref _rotateMode, value); }
      }

      protected override void OnSelfPropertyChanged(PropertyChangedEventArgs ev) {
         //LoggerSimple.Put($"OnSelfPropertyChanged: {Entity}: PropertyName={ev.PropertyName}");
         base.OnSelfPropertyChanged(ev);
         switch (ev.PropertyName) {
         case nameof(this.Size):
         case nameof(this.Padding):
            RecalcArea();
            break;
         case nameof(this.Rotate):
         case nameof(this.RotateMode):
         case nameof(this.SizeField):
            if (RotateMode == ERotateMode.SomeCells)
               RandomRotateElemenIndex();
            break;
         }
      }

      #region Dependencys
      void Dependency_CellAttribute_Area() {
         if (_cellAttr == null)
            return;
         CellAttr.Area = Area;
         if (_matrix.Any())
            foreach (var cell in Matrix)
               cell.Init();
      }
      #endregion

      protected override void OnTimer() {
         var rotateMode = RotateMode;

         double? anglePrevis = null;
         if (rotateMode == ERotateMode.SomeCells)
            anglePrevis = RotateAngle;
         base.OnTimer();

         switch (rotateMode) {
         case ERotateMode.FullMatrix:
            RotateMatrix();
            break;
         case ERotateMode.SomeCells:
            UpdateAnglesOffsets(anglePrevis.Value);
            RotateCells();
            break;
         }
      }

      #region PART ERotateMode.FullMatrix

      private void RotateMatrix() {
         var center = new PointDouble(Size.Width / 2.0 - _paddingFull.Left, Size.Height / 2.0 - _paddingFull.Top);
         foreach (var cell in Matrix) {
            cell.Init(); // restore base coords
            cell.getRegion().Points.RotateList(RotateAngle, center);
         }
      }

      #endregion

      #region PART ERotateMode.SomeCells

      private bool rotateCellAlterantive;

      protected class RotatedCellContext {
         public RotatedCellContext(int index, double angleOffset, double area) {
            this.index = index;
            this.angleOffset = angleOffset;
            this.area = area;
         }
         public readonly int index;
         public readonly double angleOffset;
         public double area;
      }

      private void RotateCells() {
         var attr = CellAttr;
         var matrix = Matrix;
         var angle = RotateAngle;
         var area = Area;

         foreach (var cntxt in RotatedElements) {
            System.Diagnostics.Debug.Assert(cntxt.angleOffset >= 0);
            var angle2 = angle - cntxt.angleOffset;
            if (angle2 < 0)
               angle2 += 360;
            System.Diagnostics.Debug.Assert(angle2 < 360);
            System.Diagnostics.Debug.Assert(angle2 >= 0);
            // (un)comment next line to look result changes...
            angle2 = Math.Sin((angle2 / 4).ToRadian()) * angle2; // accelerate / ускоряшка..

            // (un)comment next line to look result changes...
            cntxt.area = area * (1 + Math.Sin((angle2 / 2).ToRadian())); // zoom'ирую


            var cell = matrix[cntxt.index];

            cell.Init();
            var center = cell.getCenter();
            var coord = cell.getCoord();

            // modify
            attr.Area = cntxt.area;

            // rotate
            cell.Init();
            var centerNew = cell.getCenter();
            var delta = new PointDouble(center.X - centerNew.X, center.Y - centerNew.Y);
            cell.getRegion().Points
               .RotateList((((coord.x + coord.y) & 1) == 0) ? +angle2 : -angle2, rotateCellAlterantive ? center : centerNew)
               .MoveList(delta);

            // restore
            attr.Area = area;
         }

         // Z-ordering
         RotatedElements.Sort((x, y) => x.area.CompareTo(y.area));
      }

      /// <summary> list of offsets rotation angles prepared for cells </summary>
      private readonly IList<double /* angle offset */ > _prepareList = new List<double>();
      protected List<RotatedCellContext> RotatedElements { get; } = new List<RotatedCellContext>();

      private void RandomRotateElemenIndex() {
         _prepareList.Clear();
         if (RotatedElements.Any()) {
            RotatedElements.Clear();
            OnSelfPropertyChanged(nameof(this.RotatedElements));
         }

         if (!Rotate)
            return;

         // create random cells indexes  and  base rotate offset (negative)
         var len = Matrix.Count;
         var rand = new Random(Guid.NewGuid().GetHashCode());
         for (var i = 0; i < len / 4.5; ++i) {
            AddRandomToPrepareList(i == 0, rand);
         }
      }

      private void AddRandomToPrepareList(bool zero, Random rand) {
         var offset = (zero ? 0 : rand.Next(360)) + RotateAngle;
         if (offset > 360)
            offset -= 360;
         _prepareList.Add(offset);
      }

      private int NextRandomIndex(Random rand) {
         var len = Matrix.Count;
         System.Diagnostics.Debug.Assert(RotatedElements.Count < len);
         do {
            var index = rand.Next(len);
            if (RotatedElements.Any(ctxt => ctxt.index == index))
               continue;
            return index;
         } while (true);
      }

      protected void UpdateAnglesOffsets(double angleOld) {
         var angleNew = RotateAngle;
         var rotateDelta = RotateAngleDelta;
         var area = Area;
         Random rand = new Random(Guid.NewGuid().GetHashCode());

         if (_prepareList.Any()) {
            var copyList = new List<double>(_prepareList);
            for (var i = copyList.Count - 1; i >= 0; --i) {
               var angleOffset = copyList[i];
               if ((rotateDelta >= 0)
                  ? ((angleOld <= angleOffset && angleOffset <  angleNew && angleOld < angleNew) || // example: old=10   offset=15   new=20
                     (angleOld <= angleOffset && angleOffset >  angleNew && angleOld > angleNew) || // example: old=350  offset=355  new=0
                     (angleOld >  angleOffset && angleOffset <= angleNew && angleOld > angleNew))   // example: old=355  offset=0    new=5
                  : ((angleOld >= angleOffset && angleOffset >  angleNew && angleOld > angleNew) || // example: old=20   offset=15   new=10
                     (angleOld <  angleOffset && angleOffset >  angleNew && angleOld < angleNew) || // example: old=0    offset=355  new=350
                     (angleOld >= angleOffset && angleOffset <= angleNew && angleOld < angleNew)))  // example: old=5    offset=0    new=355
               {
                  _prepareList.RemoveAt(i);
                  RotatedElements.Add(new RotatedCellContext(NextRandomIndex(rand), angleOffset, area));
                  OnSelfPropertyChanged(nameof(this.RotatedElements));
               }
            }
         }

         List<RotatedCellContext> toRemove = null;
         foreach (var cntxt in RotatedElements) {
            var angle2 = angleNew - cntxt.angleOffset;
            if (angle2 < 0)
               angle2 += 360;
            System.Diagnostics.Debug.Assert(angle2 < 360);
            System.Diagnostics.Debug.Assert(angle2 >= 0);

            // prepare to next step - exclude current cell from rotate and add next random cell
            var angle3 = angle2 + rotateDelta;
            if ((angle3 >= 360) || (angle3 < 0)) {
               if (toRemove == null)
                  toRemove = new List<RotatedCellContext>();
               toRemove.Add(cntxt);
            }
         }
         if (toRemove != null) {
            foreach (var cntxt in toRemove) {
               Matrix[cntxt.index].Init(); // restore original region coords
               RotatedElements.Remove(cntxt);
               if (!RotatedElements.Any())
                  rotateCellAlterantive = !rotateCellAlterantive;
               AddRandomToPrepareList(false, rand);
            }
            OnSelfPropertyChanged(nameof(this.RotatedElements));
         }
      }

      #endregion

   }

}

File: src/src_core/fmg/core/img/AMosaicsImg.java
package fmg.core.img;

import java.beans.PropertyChangeEvent;
import java.beans.PropertyChangeListener;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

import fmg.common.geom.*;
import fmg.common.geom.util.FigureHelper;
import fmg.core.mosaic.Mosaic;
import fmg.core.mosaic.MosaicHelper;
import fmg.core.mosaic.cells.BaseCell;
import fmg.core.mosaic.cells.BaseCell.BaseAttribute;
import fmg.core.types.EMosaic;

/**
 * Abstract representable {@link fmg.core.types.EMosaic} as image
 *
 * @param <TImage> plaform specific image
 */
public abstract class AMosaicsImg<TImage>
      extends AnimatedImg<TImage>
      implements PropertyChangeListener
{
   public enum ERotateMode {
      fullMatrix,
      someCells
   }

   /** MVC: model */
   private Mosaic _mosaic;

   /** get model */
   public Mosaic getMosaic() {
      if (_mosaic == null)
         setMosaic(new Mosaic());
      return _mosaic;
   }
   /** set model */
   protected void setMosaic(Mosaic model) {
      if (_mosaic != null) {
         _mosaic.removeListener(this);
         _mosaic.close();
      }
      _mosaic = model;
      if (_mosaic != null) {
         _mosaic.addListener(this);
      }
   }

   public static final String PROPERTY_MOSAIC_TYPE      = Mosaic.PROPERTY_MOSAIC_TYPE;
   public static final String PROPERTY_SIZE_FIELD       = Mosaic.PROPERTY_SIZE_FIELD;
   public static final String PROPERTY_MATRIX           = Mosaic.PROPERTY_MATRIX;
   public static final String PROPERTY_AREA             = Mosaic.PROPERTY_AREA;
   public static final String PROPERTY_PADDING_FULL     = "PaddingFull";
   public static final String PROPERTY_ROTATE_MODE      = "RotateMode";
   public static final String PROPERTY_ROTATED_ELEMENTS = "RotatedElements";

   /** из каких фигур состоит мозаика поля */
   public EMosaic getMosaicType() { return getMosaic().getMosaicType(); }
   public void setMosaicType(EMosaic newMosaicType) { getMosaic().setMosaicType(newMosaicType); }

   public Matrisize getSizeField() { return getMosaic().getSizeField(); }
   public void setSizeField(Matrisize newSizeField) { getMosaic().setSizeField(newSizeField); }

   public BaseCell getCell(Coord coord) { return getMosaic().getMatrix().get(coord.x * getSizeField().n + coord.y); }

   /** матрица ячеек, представленная(развёрнута) в виде вектора */
   public List<BaseCell> getMatrix() { return getMosaic().getMatrix(); }

   private void recalcArea() {
      int w = getSize().width;
      int h = getSize().height;
      Bound pad = getPadding();
      SizeDouble sizeImageIn = new SizeDouble(w - pad.getLeftAndRight(), h - pad.getTopAndBottom());
      SizeDouble sizeImageOut = new SizeDouble();
      double area = MosaicHelper.findAreaBySize(getMosaicType(), getSizeField(), sizeImageIn, sizeImageOut);
      setArea(area);
      assert (w >= (sizeImageOut.width + pad.getLeftAndRight()));
      assert (h >= (sizeImageOut.height + pad.getTopAndBottom()));
      BoundDouble paddingOut = new BoundDouble(
         (w - sizeImageOut.width)/2,
         (h - sizeImageOut.height)/2,
         (w - sizeImageOut.width)/2,
         (h - sizeImageOut.height)/2);
      assert DoubleExt.hasMinDiff(sizeImageOut.width + paddingOut.getLeftAndRight(), w);
      assert DoubleExt.hasMinDiff(sizeImageOut.height + paddingOut.getTopAndBottom(), h);

      setPaddingFull(paddingOut);
   }

   public double getArea() { return getMosaic().getArea(); }
   public void setArea(double newArea) { getMosaic().setArea(newArea); }

   private BoundDouble _paddingFull;
   public BoundDouble getPaddingFull() { return _paddingFull; }
   protected void setPaddingFull(BoundDouble value) {
      if (setProperty(_paddingFull, value, PROPERTY_PADDING_FULL)) {
         invalidate();
      }
   }

   private ERotateMode _rotateMode = ERotateMode.fullMatrix;
   public ERotateMode getRotateMode() { return _rotateMode; }
   public void setRotateMode(ERotateMode value) { setProperty(_rotateMode, value, PROPERTY_ROTATE_MODE); }

   @Override
   public void propertyChange(PropertyChangeEvent ev) {
      if (ev.getSource() instanceof Mosaic)
         onMosaicPropertyChanged((Mosaic)ev.getSource(), ev);
   }

   private void onMosaicPropertyChanged(Mosaic source, PropertyChangeEvent ev) {
      String propertyName = ev.getPropertyName();
      switch (propertyName) {
      case Mosaic.PROPERTY_SIZE_FIELD:
         recalcArea();
         onSelfPropertyChanged(ev.getOldValue(), ev.getNewValue(), PROPERTY_SIZE_FIELD);
         invalidate();
         break;
      case Mosaic.PROPERTY_MOSAIC_TYPE:
         recalcArea();
         onSelfPropertyChanged(ev.getOldValue(), ev.getNewValue(), PROPERTY_MOSAIC_TYPE);
         invalidate();
         break;
      case Mosaic.PROPERTY_AREA:
         onSelfPropertyChanged(ev.getOldValue(), ev.getNewValue(), PROPERTY_AREA);
         invalidate();
         break;
      case Mosaic.PROPERTY_CELL_ATTR:
         invalidate();
         break;
      case Mosaic.PROPERTY_MATRIX:
         onSelfPropertyChanged(PROPERTY_MATRIX);
         invalidate();
         break;
      default:
         break;
      }
   }

   @Override
   protected void onSelfPropertyChanged(Object oldValue, Object newValue, String propertyName) {
      //LoggerSimple.Put("onSelfPropertyChanged: {0}: PropertyName={1}", Entity, ev.PropertyName);
      super.onSelfPropertyChanged(oldValue, newValue, propertyName);
      switch (propertyName) {
      case PROPERTY_SIZE:
      case PROPERTY_PADDING:
         recalcArea();
         break;
      case PROPERTY_ROTATE:
      case PROPERTY_ROTATE_MODE:
      case PROPERTY_SIZE_FIELD:
         if (getRotateMode() == ERotateMode.someCells)
            randomRotateElemenIndex();
         break;
      }
   }

   @Override
   protected void onTimer() {
      ERotateMode rotateMode = getRotateMode();

      Double anglePrevis = (rotateMode == ERotateMode.someCells) ? getRotateAngle() : null;
      super.onTimer();

      switch (rotateMode) {
      case fullMatrix:
         rotateMatrix();
         break;
      case someCells:
         updateAnglesOffsets(anglePrevis);
         rotateCells();
         break;
      }
   }

   /** ///////////// ================= PART {@link ERotateMode#fullMatrix} ======================= ///////////// */

   public void rotateMatrix() {
      PointDouble center = new PointDouble(getSize().width / 2.0 - _paddingFull.left, getSize().height / 2.0 - _paddingFull.top);
      for (BaseCell cell : getMatrix()) {
         cell.init(); // restore base coords

         FigureHelper.rotateCollection(cell.getRegion().getPoints(), getRotateAngle(), center);
      }
   }

   /** ///////////// ================= PART {@link ERotateMode#someCells} ======================= ///////////// */

   private boolean rotateCellAlterantive;

   protected static final class RotatedCellContext {
      public RotatedCellContext(int index, double angleOffset, double area) {
         this.index = index;
         this.angleOffset = angleOffset;
         this.area = area;
      }
      public final int index;
      public final double angleOffset;
      public double area;
   }

   /** rotate BaseCell from original Matrix with modified Region */
   protected void rotateCells() {
      BaseAttribute attr = getMosaic().getCellAttr();
      List<BaseCell> matrix = getMatrix();
      final double area = getArea();
      final double angle = getRotateAngle();

      _rotatedElements.forEach(cntxt -> {
         assert (cntxt.angleOffset >= 0);
         double angle2 = angle - cntxt.angleOffset;
         if (angle2 < 0)
            angle2 += 360;
         assert (angle2 < 360);
         assert (angle2 >= 0);
         // (un)comment next line to look result changes...
         angle2 = Math.sin(FigureHelper.toRadian((angle2 / 4))) * angle2; // accelerate / ускоряшка..

         // (un)comment next line to look result changes...
         cntxt.area = area * (1 + Math.sin(FigureHelper.toRadian(angle2 / 2))); // zoom'ирую


         BaseCell cell = matrix.get(cntxt.index);

         cell.init();
         PointDouble center = cell.getCenter();
         Coord coord = cell.getCoord();

         Mosaic m = getMosaic();
         // modify
         m.getCellAttr().removeListener(m); // disable handling Mosaic.onCellAttributePropertyChanged(where event.propertyName == BaseCell.BaseAttribute.PROPERTY_AREA)
         attr.setArea(cntxt.area);

         // rotate
         cell.init();
         PointDouble centerNew = cell.getCenter();
         PointDouble delta = new PointDouble(center.x - centerNew.x, center.y - centerNew.y);
         FigureHelper.moveCollection(FigureHelper.rotateCollection(cell.getRegion().getPoints(), (((coord.x + coord.y) & 1) == 0) ? +angle2 : -angle2, rotateCellAlterantive ? center : centerNew), delta);

         // restore
         attr.setArea(area);
         m.getCellAttr().addListener(m);
      });

      // Z-ordering
      Collections.sort(_rotatedElements, (e1, e2) -> Double.compare(e1.area, e2.area));
   }

   /** list of offsets rotation angles prepared for cells */
   private final List<Double /* angle offset */ > _prepareList = new ArrayList<>();
   protected final List<RotatedCellContext> _rotatedElements = new ArrayList<>();

   private void randomRotateElemenIndex() {
      _prepareList.clear();
      if (!_rotatedElements.isEmpty()) {
         _rotatedElements.clear();
         onSelfPropertyChanged(PROPERTY_ROTATED_ELEMENTS);
      }

      if (!isRotate())
         return;

      // create random cells indexes  and  base rotate offset (negative)
      int len = getMatrix().size();
      Random rand = ThreadLocalRandom.current();
      for (int i = 0; i < len/4.5; ++i) {
         addRandomToPrepareList(i==0, rand);
      }
   }

   private void addRandomToPrepareList(boolean zero, Random rand) {
      double offset = (zero ? 0 : rand.nextInt(360)) + getRotateAngle();
      if (offset > 360)
         offset -= 360;
      _prepareList.add(offset);
   }

   private int nextRandomIndex(Random rand) {
      int len = getMatrix().size();
      assert (_rotatedElements.size() < len);
      do {
         int index = rand.nextInt(len);
         if (_rotatedElements.stream().anyMatch(ctxt -> ctxt.index == index))
            continue;
         return index;
      } while(true);
   }

   protected void updateAnglesOffsets(double angleOld) {
      double angleNew = getRotateAngle();
      double rotateDelta = getRotateAngleDelta();
      double area = getArea();
      Random rand = ThreadLocalRandom.current();

      if (!_prepareList.isEmpty()) {
         List<Double> copyList = new ArrayList<Double>(_prepareList);
         for (int i = copyList.size()-1; i >= 0; --i) {
            double angleOffset = copyList.get(i);
            if ((rotateDelta >= 0)
               ?  ((angleOld <= angleOffset && angleOffset <  angleNew && angleOld < angleNew) || // example: old=10   offset=15   new=20
                   (angleOld <= angleOffset && angleOffset >  angleNew && angleOld > angleNew) || // example: old=350  offset=355  new=0
                   (angleOld >  angleOffset && angleOffset <= angleNew && angleOld > angleNew))   // example: old=355  offset=0    new=5
               :  ((angleOld >= angleOffset && angleOffset >  angleNew && angleOld > angleNew) || // example: old=20   offset=15   new=10
                   (angleOld <  angleOffset && angleOffset >  angleNew && angleOld < angleNew) || // example: old=0    offset=355  new=350
                   (angleOld >= angleOffset && angleOffset <= angleNew && angleOld < angleNew)))  // example: old=5    offset=0    new=355
            {
               _prepareList.remove(i);
               _rotatedElements.add(new RotatedCellContext(nextRandomIndex(rand), angleOffset, area));
               onSelfPropertyChanged(PROPERTY_ROTATED_ELEMENTS);
            }
         }
      }

      List<RotatedCellContext> toRemove = new ArrayList<>();
      _rotatedElements.forEach(cntxt -> {
         double angle2 = angleNew - cntxt.angleOffset;
         if (angle2 < 0)
            angle2 += 360;
         assert (angle2 < 360);
         assert (angle2 >= 0);

         // prepare to next step - exclude current cell from rotate and add next random cell
         double angle3 = angle2 + rotateDelta;
         if ((angle3 >= 360) || (angle3 < 0)) {
            toRemove.add(cntxt);
         }
      });
      if (!toRemove.isEmpty()) {
         List<BaseCell> matrix = getMatrix();
         toRemove.forEach(cntxt -> {
                           matrix.get(cntxt.index).init(); // restore original region coords
                           _rotatedElements.remove(cntxt);
                           if (_rotatedElements.isEmpty())
                              rotateCellAlterantive = !rotateCellAlterantive;
                           addRandomToPrepareList(false, rand);
                        });
         onSelfPropertyChanged(PROPERTY_ROTATED_ELEMENTS);
      }
   }

   @Override
   public void close() {
      super.close();
      setMosaic(null);
   }

}
